mrmustard/physics/representations.py

```diff
@@ -310,3 +315,115 @@ def plot(
+
+            fock8 = fock1.conj() # conjugate of it
+            assert fock8.array.shape == (1,5,7,8)
+    Args:
+
+    """
+
+    def __init__(self, array: Batch[Tensor]):
+        self._contract_idxs: tuple[int, ...] = ()
+        self.ansatz = ArrayAnsatz(array=array)
+
+    def from_ansatz(self, ansatz: ArrayAnsatz) -> Fock:
+        r"""Returns a Fock object from an ansatz object."""
+        return self.__class__(ansatz.array)
+
+    @property
+    def array(self) -> Batch[Tensor]:
+        r"""
+        The array from the ansatz.
+        """
+        return self.ansatz.array
+
+    def conj(self):
+        r"""
+        The conjugate of this Fock object.
+        """
+        new = self.__class__(np.conj(self.array))
+        new._contract_idxs = self._contract_idxs
+        return new
+
+    def __getitem__(self, idx: int | tuple[int, ...]) -> Fock:
+        r"""Returns a copy of self with the given indices marked for contraction."""
+        idx = (idx,) if isinstance(idx, int) else idx
+        for i in idx:
+            if i >= len(self.array.shape):
+                raise IndexError(
+                    f"Index {i} out of bounds for ansatz {self.ansatz.__class__.__qualname__} of dimension {self.ansatz.dim}."
+                )
+        new = self.__class__(self.array)
+        new._contract_idxs = idx
+        return new
+
+    def __matmul__(self, other: Fock) -> Fock:
+        r"""Implements the inner product of ansatze across the marked indices.
+
+        Batch:
+        The new Fock will hold the tensor product batch of them.
+
+        Order of index:
+        The new Fock's order is arranged as uncontracted elements in self and then other.
+        """
+        axes = [list(self._contract_idxs), list(other._contract_idxs)]
+        new_array = []
+        for a in self.array:
+            for b in other.array:
+                print("count")
+                new_array.append(math.tensordot(a, b, axes))
+        return self.__class__(array=math.astensor(new_array))
+
+    def trace(self, idxs1: tuple[int, ...], idxs2: tuple[int, ...]) -> Fock:
+        r"""Implements the partial trace over the given index pairs.
+
+        Args:
+            idxs1: indices part1 to trace over
+            idxs2: indices part2 to trace over
+
+        Returns:
+            Fock: the ansatz with the given indices traced over
+        """
+        new_array = self.array
+        for id1, id2 in zip(idxs1, idxs2):
+            list_transposed = [x for x in range(len(self.array.shape)) if x not in [id1, id2]] + [
+                id1,
+                id2,
+            ]
+            new_array = np.trace(math.transpose(self.array, list_transposed))
+        return self.__class__(array=new_array)
+
+    def reorder(self, order: tuple[int, ...] | list[int]) -> Fock:
+        r"""Reorders the indices of the array with the given order.
+        Returns a new Fock object."""
+        return self.__class__(array=math.transpose(self.array, order))
```

Comment:
Isn't this just reordering the batches? Is this what we want?

Reply:
If the user is aware of this  batch or not? If this order contains it or not. No idea. Related to batch.